### ISSUE TITLE
Support specifying a command directly

### ISSUE THREAD

**Issue by chrisbouchard** [astoff/buffer-env#17]
First, thank you for this package. It's almost exactly what I was need, and I was about to resign myself to writing something crappier by hand. 😄

I would like to suggest an option to override the &ldquo;search upward for a file with a known name&rdquo; feature, and to specify a command directly. My use-case is that I use [rtx][], a clone of [asdf][], which would have the same concerns. I think it would also apply to direnv users.

rtx already knows how to search for configuration files, and it recognizes [a variety of different file names][rtx-config]&mdash;actually an unbounded list of file names, because plugins can add their own. I'd rather not try to replicate all the different file names possible, and it's not uncommon for multiple config files could occur in the same directory. I would always want to run one command, `rtx exec -- env -0`, regardless what directory the buffer is in or what other files exist.

I realize I could create a `.envrc` file in my home directory that uses `source $(rtx env)` to set the environment, and let `buffer-env-update` find it&mdash;~~and I think that will be my work-around for now~~ (Edit: I don't think this would actually work. See the comment below.)&mdash;but it seems wasteful to search all the way up to my home directory, just to do it again inside rtx. Also, when editing files outside my home directory, I would still want to use the environment from `rtx`. (It has global configuration.) It also feels a little silly to set the environment via `bash` when rtx already provides it via `rtx exec`&mdash;and probably more efficiently, since it's a compiled program.

I don't know what the best interface would be to support this option. I can think of a few alternatives:

1. Allow `buffer-env-script-name` and the file argument to `buffer-env-update` to be `nil`, and allow `nil` as a key in `buffer-env-commands`.
    * Pros: Uses the existing interface. Still allows files to be sourced interactively.
    * Cons: Not obvious what `nil` indicates. Only allows one command to be defined.

2. Add a `buffer-env-treat-as-command` flag, which would cause `buffer-env-script-name` and the file argument to `buffer-env-update` to be interpreted instead as literal commands.
    * Pros: Allows multiple commands and interactive commands.
    * Cons: It would be confusing that `buffer-env-script-name` might be a script. No way to interactively source files if flag is set.

3. Deprecate `buffer-env-script-name` in favor of a new `buffer-env-sources` list accepting a mix of strings and type-string pairs, where type is one of `file` or `command`, and bare strings are treated as files. Commands would be run unconditionally; files would be searched and evaluated as normal. The file argument to `buffer-env-update` would accept a string or type-string pair, or else might learn an optional `:type` keyword argument, or both.
    * Pros: Most flexible, I think. Allows multiple commands. Allows sourcing file and running commands interactively. Leaves space for more types in the future.
    * Cons: Bigger change. Possibly more complex to implement. More to explain.

[asdf]: https://asdf-vm.com/
[rtx]: https://github.com/jdxcode/rtx
[rtx-config]: https://github.com/jdxcode/rtx#configuration

**Comment by astoff** [comment]
I haven't tested, but can't you set `buffer-env-script-name` to `"."` and adjust `buffer-env-commands` accordingly?

However, because `buffer-env` caches the env variables computed for each script, I wouldn't recommend the above anyway; rather, I'd recommend to still look up the parent directories for some marker file in order to determine the project root. IIUC, in the rtx case it would be .tool-versions or .rtx.toml (but you could even use `.git` if that's an assumption you can make of all your projects).

So, in short, set `buffer-env-script-name` to `".rtx.toml"` and add `(".rtx.toml" . "rtx exec -- env -0")` to `buffer-env-commands`. (This second bit is something I could add to the package if you think that's good enough for rtx users.)

**Comment by chrisbouchard** [comment]
I haven't considered trying `.`, but I will give it a try. (It's kind of like alternative 1 for specifying commands via `nil` in my previous post.) But yeah, I think buffer-env would cache the directory entry, so I guess the only way to refresh environment variables would be to update the directory's modification time, which is less than ideal. I hadn't considered the cache when I mentioned `~/.envrc` as a work-around, so don't think that would actually work. I think using `.git` would have the same problem as `.`. Even project files like `.tool-versions` wouldn't necessarily interact well with the cache, because they might only partially override the config from a parent `.tool-versions` file, and there's the global rtx config in `~/.config/rtx/config.toml`.

I suppose I'd add to my suggestion that caching and authorization should only apply to files and not to commands, because commands aren't directory-local. I guess a command *could* be directory-local if a user configured a command like `./script.sh`, but that would also error out frequently for not existing, and they should probably have it as a file source. For what it's worth, rtx and direnv both already have config file authorization mechanisms.

I feel like decoupling the &ldquo;buffer-local environment&rdquo; feature and the &ldquo;file loading, caching, and authorizing&rdquo; feature of this package would be really handy. Both features are really useful, but being able to use them independently would be much more flexible. :+1:

As for adding `.rtx.toml` or `.tool-versions`, I'd also have to add `.nvmrc`, `.node-version`, `.ruby-version`, etc. It's pretty common for projects to use them, since they're language-specific rather than tool-specific, and one of the features of rtx and asdf is that they *Just Work&trade;* even if there's no `.rtx.toml` or `.tool-versions`. I'd rather not have to `touch` an unnecessary file to make environment variables work&mdash;I will forget every time. :smile: A project could even have `.node-version` *and* `.ruby-version`, if it uses both languages, which would result in double-loading the environment.

**Comment by astoff** [comment]
> I hadn't considered the cache when I mentioned ~/.envrc as a work-around, so don't think that would actually work.

It works in the sense that you can `touch` that file manually after changing the configuration.

> I feel like decoupling the “buffer-local environment” feature and the “file loading, caching, and authorizing” feature of this package would be really handy.

From my perspective, the "buffer-local environment" feature isn't really a feature of this package: I just set `process-environment` buffer-locally, which is trivial. (Okay, I also update `exec-path` accordingly, so there's an epsilon of work done on the buffer-env side.)

So the actual purpose of buffer-env (“file loading, caching, and authorizing”, as you describe it) overlaps almost completely with rtx. The difference is that rtx has a much more complex logic to determine what to "load". If you really want to rely on that logic, you might be better off with "writing something crappier by hand" (which doesn't need to be crappy). A starting point would be

```elisp
(defun rxt ()
  (ignore-errors
    (setq-local process-environment (process-lines "rtx exec -- env"))))
(add-hook 'hack-local-variables-hook 'rtx)
```

> As for adding .rtx.toml or .tool-versions, I'd also have to add .nvmrc, .node-version, .ruby-version, etc

You could do that in your config, but I couldn't add it as a default, since I can't assume that all Node users are rtx users.

By the way, I'm not sure you're describing what you personally need or what you believe a generic user needs. Do you deal with so many different kinds of project that you couldn't find 2 or 3 file names and corresponding commands that mostly cover your needs?

> and maybe even an additional .tool-versions for a runtime or LSP

I would be very annoyed if a project required a specific environment management tool :-)

I don't know how often you start/clone a new project, but to me the goal is not to make that a zero effort step (which is impossible — you have at least to create/clone a directory and also mark some configuration file as safe to execute). Rather, my goal is to make "work on a project" or "switch projects" be zero effort. Therefore, it seems to me that creating a possibly empty `.rtx.toml` is about as easy as it gets (an empty `.rtx.toml` works fine if the project has "legacy" config files, right?).

**Comment by astoff** [comment]
I'll close this due to inactivity, let me know if you still have ideas.
